### DISCUSSION TITLE
How to grab gradients?

### DISCUSSION THREAD

**Discussion by patins1** [deepjavalibrary/djl#1503] [General]
Hi, 

During training, at the onEpoch event, I want to read the gradients for certain model parameters, but they are all zero already.
I realized, they become zero by a call to ParameterStore.updateAllParameters() during a training step (I debugged it down to a call of NDArrayEx.adamUpdate()). It surely makes sense to put the parameters to zero again after a training step, however it doesn't let me log the gradients for analysis reasons, is there a better approach to achieve that?

**Comment by zachgk** [comment]
I assume you are using `EasyTrain` so correct me if I am wrong, but it may not work entirely for you. There aren't any TrainingListener events that occur mid-batch right now. I recommend forking the whole `EasyTrain` class. Then, you can access the gradients inside `trainBatch` before they are cleared

**Comment by patins1** [comment]
Thanks, I found the event to get grab the gradient eventually: onTrainingBatch
